src/commands.rs

```diff
@@ -65,11 +65,11 @@ fn format_filter(config: &config::Config) -> String {
   }
 }
 
-/// Find a Nitrokey device that matches the given requirements
-fn find_device(config: &config::Config) -> anyhow::Result<nitrokey::DeviceInfo> {
+/// Find all Nitrokey devices that match the given requirements
+fn find_devices(config: &config::Config) -> anyhow::Result<Vec<nitrokey::DeviceInfo>> {
```

Comment:
Hm. I'd not want to introduce unnecessary allocations where they can be avoided (and yes, I do realize `nitrokey-rs` already allocates a vector). We could remove the need for a `Vec` here with the following patch:
```diff
--- src/commands.rs
+++ src/commands.rs
@@ -65,23 +65,27 @@ fn format_filter(config: &config::Config) -> String {
   }
 }

-/// Find all Nitrokey devices that match the given requirements
-fn find_devices(config: &config::Config) -> anyhow::Result<Vec<nitrokey::DeviceInfo>> {
-  let devices = nitrokey::list_devices().context("Failed to enumerate Nitrokey devices")?;
+/// Filter out all Nitrokey devices that match the given requirements
+fn filter_devices<'c, I>(
+  devices: I,
+  config: &'c config::Config,
+) -> impl Iterator<Item = nitrokey::DeviceInfo> + 'c
+where
+  I: ExactSizeIterator<Item = nitrokey::DeviceInfo> + 'c,
+{
   let nkmodel = config.model.map(nitrokey::Model::from);
-  let matching_devices = devices
-    .into_iter()
-    .filter(|device| nkmodel.is_none() || device.model == nkmodel)
-    .filter(|device| {
+  devices
+    .filter(move |device| nkmodel.is_none() || device.model == nkmodel)
+    .filter(move |device| {
       config.serial_numbers.is_empty()
         || device
           .serial_number
           .map(|sn| config.serial_numbers.contains(&sn))
           .unwrap_or_default()
     })
-    .filter(|device| config.usb_path.is_none() || config.usb_path.as_ref() == Some(&device.path))
-    .collect();
-  Ok(matching_devices)
+    .filter(move |device| {
+      config.usb_path.is_none() || config.usb_path.as_ref() == Some(&device.path)
+    })
 }

 /// Connect to a Nitrokey device that matches the given requirements
@@ -89,18 +93,20 @@ fn connect<'mgr>(
   manager: &'mgr mut nitrokey::Manager,
   config: &config::Config,
 ) -> anyhow::Result<nitrokey::DeviceWrapper<'mgr>> {
-  let device_infos = find_devices(config)?;
+  let devices = nitrokey::list_devices().context("Failed to enumerate Nitrokey devices")?;
+  let mut device_infos = filter_devices(devices.into_iter(), config);
+
+  let device = device_infos
+    .next()
+    .with_context(|| format!("Nitrokey device not found{}", format_filter(config)))?;

   anyhow::ensure!(
-    device_infos.len() < 2,
+    device_infos.next().is_none(),
     "Multiple Nitrokey devices found{}.  Use the --model, --serial-number, and --usb-path options \
     to select one",
     format_filter(config)
   );

-  let device = device_infos
-    .first()
-    .with_context(|| format!("Nitrokey device not found{}", format_filter(config)))?;
   manager.connect_path(device.path.deref()).with_context(|| {
     format!(
       "Failed to connect to Nitrokey device at path {}",
```

What do you think?

Reply:
I thought about this, but I decided that I’d be willing to pay for easier code with one vector allocation.  But I can change that if you want to.